README.md
# ZenWeb Metric module

zenweb 健康检测模块，用于生产环境的实例运行信息收集

[ZenWeb](https://www.npmjs.com/package/zenweb)

## 快速使用

安装模块
```bash
$ npm i @zenweb/metric
```

配置模块： 编辑项目启动文件
```ts
// 导入模块
import metric from "@zenweb/metric";
// 安装模块
app.setup(metric(/* 可选配置项 */));
```

## 配置项
| 项 | 值类型 | 默认值 | 说明 |
|----|-------|-------|-----|
| logDir | `string` | `process.env.ZENWEB_METRIC_LOG_DIR \|\| os.tmpdir()` | 日志输出目录 |
| logInterval | `number` | `process.env.ZENWEB_METRIC_LOG_INTERVAL \|\| 10` | 日志输出间隔(秒) |
| apdexSatisfied | `number` | `100` | apdex 满意值(毫秒内) |

## Changelog

### 3.2.0
- null 检查
- 不再依赖 meta 模块


CHANGELOG.md
# Changelog

## [3.3.0] - 2023-5-30
- 使用 zenlog.js 输出日志

## 3.2.1
- 删除无用选项: enableProcessTitle

## 3.2.0
- null 检查
- 不再依赖 meta 模块


src/index.ts
import * as os from 'os';
import { SetupFunction } from '@zenweb/core';
import { FileBufferStream, FileBufferStreamOption, RootLogger } from 'zenlog.js';

export interface MetricOption extends FileBufferStreamOption {
  /**
   * 日志输出目录
   * @default process.env.LOG_DIR || '/tmp'
   */
  dir?: string;

  /**
   * 日志文件名
   * @default 'zenweb-metric.{yyyy}-{mm}-{dd}.log'
   */
  filename?: string;

  /**
   * 取样间隔(毫秒)
   * @default process.env.ZENWEB_METRIC_SAMPLE_INTERVAL || 10000
   */
  sampleInterval?: number;

  /**
   * apdex 满意值(毫秒内)
   * @default process.env.ZENWEB_METRIC_APDEX_SATISFIED || 100
   */
  apdexSatisfied?: number;
}

const defaultOption: MetricOption = {
  dir: process.env.LOG_DIR,
  filename: 'zenweb-metric.{yyyy}-{mm}-{dd}.log',
  sampleInterval: parseInt(process.env.ZENWEB_METRIC_SAMPLE_INTERVAL || ''),
  apdexSatisfied: parseInt(process.env.ZENWEB_METRIC_APDEX_SATISFIED || ''),
};

export default function setup(opt?: MetricOption): SetupFunction {
  const option = Object.assign({}, defaultOption, opt);
  return function metric(setup) {
    setup.debug('option: %o', option);

    const logger = new RootLogger({
      name: setup.core.name,
      instance: `${os.hostname()}-${process.pid}`,
    });
    logger.addStream(new FileBufferStream(option));

    const cpuCount = os.cpus().length;
    const sampleInterval = option.sampleInterval || 10000;
  
    let lastSampleCpuUsage = process.cpuUsage();
    let lastSampleTime = Date.now();
    let lastRequests = 0;
    let lastRequestsElapsed = 0;
    let lastApdexTolerates = 0;
    let metricRequests = 0;
    let metricRequestsElapsed = 0;
    let metricApdexTolerates = 0;

    function collect() {
      // collect
      const elapsedTime = Date.now() - lastSampleTime;
      const cpuUsage = process.cpuUsage(lastSampleCpuUsage);
      const requests = metricRequests - lastRequests;
      const requests_elapsed = metricRequestsElapsed - lastRequestsElapsed;
      const apdexTolerates = metricApdexTolerates - lastApdexTolerates;
  
      // write
      const mem = process.memoryUsage();
      const data: { [key: string]: any } = {
        timestamp: Math.round(Date.now() / 1000),
        cpu_percentage: (cpuUsage.user + cpuUsage.system) / 1000 / elapsedTime,
        event_delay: Math.max(0, elapsedTime - sampleInterval),
        mem_rss: mem.rss,
        mem_heap_total: mem.heapTotal,
        mem_heap_used: mem.heapUsed,
        mem_external: mem.external,
        mem_array_buffers: mem.arrayBuffers,
        mem_os_free: os.freemem(),
        load_percentage: os.loadavg()[0] / cpuCount,
        active_handles: proceeActiveHandles().length,
      };
      if (requests > 0) {
        data.requests = requests;
        data.requests_elapsed = requests_elapsed;
        data.qps = requests / elapsedTime * 1000;
        data.apdex = ((requests - apdexTolerates) + apdexTolerates * 0.5) / requests;
      }

      // reset
      lastSampleCpuUsage = process.cpuUsage();
      lastSampleTime = Date.now();
      lastRequests = metricRequests;
      lastRequestsElapsed = metricRequestsElapsed;
      lastApdexTolerates = metricApdexTolerates;
  
      setup.debug(data);

      // 写入文件
      logger.record(data);
    }
  
    const timer = setInterval(() => collect(), sampleInterval);

    setup.middleware(async function metricMiddleware(ctx, next) {
      metricRequests++;
      const startTime = Date.now();
      try {
        await next();
      } finally {
        const elapsed = Date.now() - startTime;
        metricRequestsElapsed += elapsed;
        if (elapsed > (option.apdexSatisfied || 100)) {
          metricApdexTolerates++;
        }
      }
    });

    setup.destroy(async () => {
      clearInterval(timer);
      await logger.close();
    });
  }
}

/**
 * 修正 typescript 编译错误, _getActiveHandles 为隐藏方法
 */
function proceeActiveHandles(): any[] {
  const p:any = process;
  return p._getActiveHandles();
}


index.js
'use strict';

const fs = require('fs');
const os = require('os');
const path = require('path');
const debug = require('debug')('zenweb:metric');

function hrtime2ms(time) {
  return time[0] * 1e3 + time[1] * 1e-6;
}

function getMetricMiddleware(core, options) {
  return async function metricMiddleware(ctx, next) {
    core._metricRequests++;
    ctx._metricStartTime = process.hrtime();
    try {
      await next();
    } finally {
      const elapsed = hrtime2ms(process.hrtime(ctx._metricStartTime));
      core._metricRequestsElapsed += elapsed;
      if (elapsed > options.apdexSatisfied) {
        core._metricApdexTolerates++;
      }
    }
  };
}

/**
 * @param {import('@zenweb/core').Core} core 
 * @param {*} [options]
 */
function setup(core, options) {
  options = Object.assign({
    name: process.env.npm_package_name || os.hostname(),
    logDir: process.env.ZENWEB_METRIC_LOG_DIR || os.tmpdir(),
    logInterval: parseInt(process.env.ZENWEB_METRIC_LOG_INTERVAL) || 10,
    apdexSatisfied: 100,
    enableProcessTitle: false, // 是否显示在进程标题中
  }, options);
  debug('options: %o', options);

  // if (options.enableProcessTitle) {
  //   const commandTitleLength = `zenweb: ${options.name} [00000] 100%`.length;
  //   if (process.title.length < commandTitleLength) {
  //     console.warn(`zenweb-metric: enableProcessTitle=true The command line length cannot be less than ${commandTitleLength}`);
  //   }
  // }

  core._metricRequests = 0;
  core._metricRequestsElapsed = 0;
  core._metricApdexTolerates = 0;

  core.koa.use(getMetricMiddleware(core, options));

  if (!fs.existsSync(options.logDir)) {
    fs.mkdirSync(options.logDir);
  }

  const instance = `${os.hostname()}-${process.pid}`;
  const cpuCount = os.cpus().length;
  const sampleInterval = options.logInterval * 1000;

  let lastSampleCpuUsage = process.cpuUsage();
  let lastSampleTime = process.hrtime();
  let lastRequests = 0;
  let lastRequestsElapsed = 0;
  let lastApdexTolerates = 0;

  setInterval(() => {
    // collect
    const elapsedTime = hrtime2ms(process.hrtime(lastSampleTime));
    const cpuUsage = process.cpuUsage(lastSampleCpuUsage);
    const requests = core._metricRequests - lastRequests;
    const requests_elapsed = core._metricRequestsElapsed - lastRequestsElapsed;
    const apdexTolerates = core._metricApdexTolerates - lastApdexTolerates;

    // write
    const now = new Date();
    const m = now.getMonth() + 1;
    const d = now.getDate();
    const ymd = `${now.getFullYear()}-${m < 10 ? '0' : ''}${m}-${d < 10 ? '0' : ''}${d}`;
    const filename = path.join(options.logDir, `zenweb-metric.${ymd}.log`);
    const mem = process.memoryUsage();
    const data = {
      name: options.name,
      instance,
      timestamp: Math.round(now / 1000),
      cpu_percentage: (cpuUsage.user + cpuUsage.system) / 1000 / elapsedTime,
      event_delay: Math.max(0, elapsedTime - sampleInterval),
      mem_rss: mem.rss,
      // mem_heap_total: mem.heapTotal,
      // mem_heap_used: mem.heapUsed,
      // mem_external: mem.external,
      // mem_array_buffers: mem.arrayBuffers,
      // mem_os_free: os.freemem(),
      load_percentage: os.loadavg()[0] / cpuCount,
      active_handles: process._getActiveHandles().length,
    };
    if (requests > 0) {
      data.requests = requests;
      data.requests_elapsed = requests_elapsed;
      data.qps = requests / elapsedTime * 1000;
      data.apdex = ((requests - apdexTolerates) + apdexTolerates * 0.5) / requests;
    }

    debug('write log: %s, %o', filename, data);

    if (options.enableProcessTitle) {
      process.title = `zenweb: ${data.name} [${data.active_handles}] ${data.qps||'-'}/QPS ${data.apdex > 0 ? Math.round(data.apdex * 100) : '-'}%`;
    }
    fs.appendFile(filename, JSON.stringify(data) + '\n', 'utf-8', err => {
      if (err) {
        core.log.error('zenweb:metric write log error: %s', err.message);
      }
    });

    // reset
    lastSampleCpuUsage = process.cpuUsage();
    lastSampleTime = process.hrtime();
    lastRequests = core._metricRequests;
    lastRequestsElapsed = core._metricRequestsElapsed;
    lastApdexTolerates = core._metricApdexTolerates;
  }, sampleInterval);
}

module.exports = {
  setup,
};
